flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReaderImplTest.java

```diff
@@ -0,0 +1,205 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toMap;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.HAS_MORE_DATA;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.NO_MORE_DATA;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * {@link ChannelStateReaderImpl} test.
+ */
+public class ChannelStateReaderImplTest {
+
+	private static final InputChannelInfo CHANNEL = new InputChannelInfo(1, 2);
+	private static final byte[] DATA = generateData(10);
+	private ChannelStateReaderImpl reader;
+
+	@Before
+	public void init() {
+		reader = getReader(CHANNEL, DATA);
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		reader.close();
+	}
+
+	@Test
+	public void testDifferentBufferSizes() throws Exception {
+		for (int bufferSize = 1; bufferSize < 2 * DATA.length; bufferSize++) {
+			try (ChannelStateReaderImpl reader = getReader(CHANNEL, DATA)) { // re-create reader to re-read the same channel
+				readAndVerify(bufferSize, CHANNEL, DATA, reader);
+			}
+		}
+	}
+
+	@Test
+	public void testWithOffsets() throws IOException {
+		Map<InputChannelStateHandle, byte[]> handlesAndBytes = generateHandlesWithBytes(10, 20);
+		ChannelStateReader reader = new ChannelStateReaderImpl(taskStateSnapshot(handlesAndBytes.keySet()), new ChannelStateSerializerImpl());
+		for (Map.Entry<InputChannelStateHandle, byte[]> e : handlesAndBytes.entrySet()) {
+			readAndVerify(42, e.getKey().getInfo(), e.getValue(), reader);
+		}
+	}
+
+	@Test(expected = Exception.class)
+	public void testReadOnlyOnce() throws IOException {
+		reader.readInputData(CHANNEL, getBuffer(DATA.length));
+		reader.readInputData(CHANNEL, getBuffer(DATA.length));
+	}
+
+	@Test(expected = Exception.class)
+	public void testReadClosed() throws Exception {
+		reader.close();
+		reader.readInputData(CHANNEL, getBuffer(DATA.length));
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testReadWrongChannelState() throws IOException {
+		InputChannelInfo wrongChannel = new InputChannelInfo(CHANNEL.getGateIdx() + 1, CHANNEL.getInputChannelIdx() + 1);
+		reader.readInputData(wrongChannel, getBuffer(DATA.length));
+	}
+
+	private TaskStateSnapshot taskStateSnapshot(Collection<InputChannelStateHandle> inputChannelStateHandles) {
+		return new TaskStateSnapshot(Collections.singletonMap(
+			new OperatorID(),
+			new OperatorSubtaskState(
+				StateObjectCollection.empty(),
+				StateObjectCollection.empty(),
+				StateObjectCollection.empty(),
+				StateObjectCollection.empty(),
+				new StateObjectCollection<>(inputChannelStateHandles),
+				StateObjectCollection.empty()
+			)));
+	}
+
+	private static byte[] generateData(int len) {
```

Comment:
nit: this can be reused in other places, like `ChannelStateSerializerImplTest#getData(int len)`